pkg/statemetrics/mr_state_metrics.go

```diff
@@ -0,0 +1,141 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package statemetrics
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/fieldpath"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+)
+
+// A MRStateRecorderOption configures a MRStateRecorder.
+type MRStateRecorderOption func(*MRStateRecorder)
+
+// A MRStateRecorder records the state of managed resources.
+type MRStateRecorder struct {
+	client    client.Client
+	log       logging.Logger
+	frequency time.Duration
+
+	mrExists *prometheus.GaugeVec
+	mrReady  *prometheus.GaugeVec
+	mrSynced *prometheus.GaugeVec
+}
+
+// NewMRStateRecorder returns a new MRStateRecorder which records the state of managed resources.
+func NewMRStateRecorder(client client.Client, log logging.Logger, frequency time.Duration) *MRStateRecorder {
+	return &MRStateRecorder{
+		client:    client,
+		log:       log,
+		frequency: frequency,
+
+		mrExists: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Subsystem: subSystem,
+			Name:      "managed_resource_exists",
+			Help:      "The number of managed resources that exist",
+		}, []string{"gvk"}),
+		mrReady: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Subsystem: subSystem,
+			Name:      "managed_resource_ready",
+			Help:      "The number of managed resources in Ready=True state",
+		}, []string{"gvk"}),
+		mrSynced: prometheus.NewGaugeVec(prometheus.GaugeOpts{
+			Subsystem: subSystem,
+			Name:      "managed_resource_synced",
+			Help:      "The number of managed resources in Synced=True state",
+		}, []string{"gvk"}),
+	}
+}
+
+// Describe sends the super-set of all possible descriptors of metrics
+// collected by this Collector to the provided channel and returns once
+// the last descriptor has been sent.
+func (r *MRStateRecorder) Describe(ch chan<- *prometheus.Desc) {
+	r.mrExists.Describe(ch)
+	r.mrReady.Describe(ch)
+	r.mrSynced.Describe(ch)
+}
+
+// Collect is called by the Prometheus registry when collecting
+// metrics. The implementation sends each collected metric via the
+// provided channel and returns once the last metric has been sent.
+func (r *MRStateRecorder) Collect(ch chan<- prometheus.Metric) {
+	r.mrExists.Collect(ch)
+	r.mrReady.Collect(ch)
+	r.mrSynced.Collect(ch)
+}
+
+// Record records the state of managed resources.
+func (r *MRStateRecorder) Record(ctx context.Context, gvk schema.GroupVersionKind) {
+	l := &unstructured.UnstructuredList{}
+	l.SetGroupVersionKind(gvk)
+	err := r.client.List(ctx, l)
```

Comment:
also here: pass a cache please to make sure this does not hit the apiserver. This can be a very heavy operation, listing thousands of MRs.

Reply:
Would it be sufficient to accept `client.Reader` and add a comment letting the caller know it must be cache-backed?

I don't feel strongly - mostly wondering if there's some way to do this up at the controller-runtime level of abstraction without moving down into client-go things. Would be nice, but not critical.

Reply:
What do you think about passing the following option to the `NewManager` function in provider's `main.go`?
```
mgr, err := ctrl.NewManager(ratelimiter.LimitRESTConfig(cfg, *maxReconcileRate), ctrl.Options{
	Client: client.Options{Cache: &client.CacheOptions{Unstructured: true}},
	...
})
```
This makes `shouldByPassCache` method to return false in the [List](https://github.com/kubernetes-sigs/controller-runtime/blob/7ff5801735fbeabed179125571bfd9e6e73068d3/pkg/client/client.go#L374) call.

Reply:
I no longer use unstructured list, so we'll use client's cache.